src/main/java/io/cdap/plugin/googleads/source/batch/GoogleAdsRecordReader.java

```
@@ -0,0 +1,187 @@
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * RecordReader implementation, which reads report instances from Google adWords using
+ * googleads-java-lib.
+ */
+public class GoogleAdsRecordReader extends RecordReader<NullWritable, StructuredRecord> {
+
+  private static final Gson gson = new GsonBuilder().create();
+  private List<StructuredRecord> reportStructure;
+  private String report;
+  private String[] heads;
+  private Iterator<StructuredRecord> iterator;
+  private StructuredRecord currentValue;
+
+
+  @Override
+  public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext) throws IOException, InterruptedException {
+
+    Configuration conf = taskAttemptContext.getConfiguration();
+    String configJson = conf.get(GoogleAdsInputFormatProvider.PROPERTY_CONFIG_JSON);
+    GoogleAdsBatchSourceConfig googleAdsBatchSourceConfig = gson.fromJson(configJson, GoogleAdsBatchSourceConfig.class);
+
+    downloadReport(googleAdsBatchSourceConfig);
+    heads = googleAdsBatchSourceConfig.getReport_fields();
+    buildReportStructure(googleAdsBatchSourceConfig.getSchema());
+
+  }
+
+  private void downloadReport(GoogleAdsBatchSourceConfig googleAdsBatchSourceConfig) throws IOException {
+    AdWordsSession session = null;
+    Credential credential = null;
+    try {
+      credential = new OfflineCredentials.Builder()
+        .forApi(OfflineCredentials.Api.ADWORDS)
+        .withClientSecrets("", "")//todo remove hardcod
+        .withRefreshToken("")//todo remove hardcod
+        .build()
+        .generateCredential();
+    } catch (OAuthException e) {
+      e.printStackTrace();
```

Don't ignore exceptions.